MPU Flash write

fvincenzo opened the pull request:
This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/54. Please do not merge directly.

The new patches are the last 3.

Commit:
arm: soc: nxp k6x: Add Allow write on Flash

This patch adds the allow flash write CONFIG option to the NXP MPU
configuration in privileged mode.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

superna9999 committed:
arm: stm32l4: Add configuration and dts for STM32L496

Add configuration and dts for the STM32L496 SoC STM32L4 variant.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
boards: arm: Add support for STM32L496G Discovery board

Add configuration, pinmux, dts and documentation for the STM32L496G
Discovery board based on the STM32L496AG SoC.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
ext: stm32cube: update stm32l4xx cube version

Update Cube version for STM32L4XX family
from version: V1.6.0
to version: V1.8.0

Add support for following soc:
stm32l496xx
stm32l4a6xx

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

Commit:
arm: core: mpu: Add Allow write on Flash

This patch adds the allow flash write CONFIG option to the ARM MPU
configuration in privileged mode.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
samples: mpu_test: Add Allow Flash Write config

This patch adds a configuration file that tests Flash writes with MPU
enabled.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>

superna9999 committed:
arm: Add dts for STM32F407

Add missing dts for STM32F407.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
arm: Add support for STM32F4DISCOVERY Board

Add configuration, documentation, flash script, pinmux and dts for the
STM32F4DISCOVERY board.

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

galak committed:
arm: dts: cleanup HAS_DTS from defconfig

Since we now select HAS_DTS from the cortex-m Kconfig we don't need it
in 96b_carbon_nrf51_defconfig.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
serial: dts: remove !HAS_DTS related Kconfig

As we now only support DTS for ARM based SoCs we can remove any
associated !HAS_DTS bits in Kconfig.  The Nordic NRF5 serial driver and
the ARM CMSDK APB serial drivers had Kconfig bits related to !HAS_DTS
builds.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak reviewed (changes requested):
Can you push to arm branch instead

mbolivar commented on arch/arm/core/cortex_m/mpu/Kconfig:
```diff
@@ -38,3 +38,10 @@ config NXP_MPU
+config MPU_ALLOW_FLASH_WRITE
+	bool "Add MPU access to write to flash"
+	depends on NXP_MPU
+	default n
```
Thanks a lot for submitting this! It will really help us.

What do people think about making this option `default y if FLASH` or so?

That seems less surprising; by default, asking for both the flash driver and the MPU driver would result in a config with fewer pointy edges. User effort would only be needed to get the flash driver and to allow flash reads, but not writes. @galak pointed out that this is a confusing configuration whose utility is a bit unclear, so perhaps that setup should take more effort to create.

fvincenzo replied:
@mbolivar I agree with the approach.

MaureenHelm replied:
In theory, yes I agree. But I wonder if there are security implications? @agross-linaro

agross-linaro replied:
Default options should not make things less secure.  I'd rather the user have to jump through 1 more hoop if they want to specifically allow this.  That's my two cents.

mbolivar replied:
It seems you're saying that asking for the MPU driver means getting a maximally locked down configuration (that can still run code), and any relaxations should require additional knobs. Is that right?

If so, it's a different point on the security / convenience curve, but that seems reasonable to me.

agross-linaro replied:
To be a little more precise:
If MPU is enabled (or future task isolation, etc), the default for all added options should be to maintain the current level of security.  The default should never relax security.

mbolivar replied:
I think I hear what you're saying.

Of course, by enabling a driver, the user expects all of the routines it declares to work, but there is potentially some conflict between the flash and MPU drivers here, depending on what you mean by "level of security" (e.g., attacks on flash involving privileged code execution aren't stopped by MPUs, since they can disable the MPU before writing flash).

It seemed like resolving that in favor of security or least-surprise was the decision.

But @galak, I think you're saying:

- add a flash config option, FLASH_DISABLE_WRITE_SUPPORT, which is default y
- don't declare flash_write() or flash_erase() in flash.h unless FLASH_DISABLE_WRITE_SUPPORT is unset
- MPU disables flash writes if and only if FLASH_DISABLE_WRITE_SUPPORT=y

That will cause churn for existing flash users, but seems like it addresses all the concerns: users get everything they ask for, and MPU disables flash writes unless explicitly requested not to.

What do you think, @agross-linaro?

agross-linaro replied:
This sounds fine, my only complaint is the naming.  I don't like options where you have DISABLE and say Y.  It's negative logic.  I'd rather have a ENABLE=y, so FLASH_ENABLE_WRITE_SUPPORT.

mbolivar replied:
SGTM.

d3zd3z replied:
Sorry to chime in late here, but it sounds like we're going for a seemingly more secure, and definitely more surprising configuration.

Preventing writes to flash isn't really all that much of a security improvement. In order to actually have secure boot, the bootloader has to be protected by hardware. MPU is insufficient, because it is easy to break before it is set up and then just replace the main code.

Once the bootloader is running, there is no security issue with being able to write to the code space. The images must be signed in order to be bootable.

In addition, apps using the bootloader are going to want to do OTA, which will require flash writing.

I really don't see any scenario where the flash driver would be wanted, but writes not allowed to the flash device.

d3zd3z replied:
Also, at least the k6x devices mentioned in the commit text have the ability to write protect flash regions, which is what will be needed to actually provide secure boot in a meaningful way.

d3zd3z replied:
One last point, it seems that on NXP, just probing the flash driver requires write access, whether the app tries to write or not. It kind of makes having this selectable a moot point. If you enable the flash driver, with writes disabled, it faults before main is even reached.

fvincenzo commented:
@galak pushed against arm branch.

dbkinder commented on samples/mpu_test/README.rst:
```diff
@@ -46,6 +46,13 @@ single thread: :file:`prj_single.conf`:
    $ make BOARD=v2m_beetle CONF_FILE=prj_single.conf run
 
+To build the version that allows flash write, use the supplied configuration
+file: :file:`prj_allow_flash_write.conf`:
```
no such file  (there are prj_allow_flash_write_mcux.conf and prj_allow_flash_write_stm32.conf).  Maybe use the same notation you use in the example below prj_allow_flash_write_<x>.conf and tell them what the <x> means.

fvincenzo replied:
Ok I will change it. Thanks

dbkinder reviewed (changes requested):
recommend fixing :file: reference to prj_allow_flash_write.conf

MaureenHelm commented on include/arch/arm/cortex_m/mpu/nxp_mpu.h:
```diff
@@ -26,8 +26,16 @@
+#if defined(CONFIG_MPU_ALLOW_FLASH_WRITE)
+/* Required to initialize the flash driver */
+#define REGION_RAM_ATTR (MPU_REGION_READ | MPU_REGION_WRITE | \
+			 MPU_REGION_EXEC |  MPU_REGION_SU)
```
I think we should bring more attention to the fact that not only are we enabling flash write, but also execution from ram. Can you mention this explicitly in the commit message? Is there a better config name to indicate this?

fvincenzo replied:
Ok I will make it more clear. Thanks for pointing it out.

agross-linaro replied:
I agree with Maureen.  I'd think about a more descriptive config name if that is easy to deal with.  Otherwise just deal with it in the commit msg.

galak committed:
arm: Fixup sanity check list of boards

The stm32f4discovery was incorrectly named and should have been
stm32f4_disco.  Also added 96b_carbon_nrf51 that was missing from the
list.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
Do we have an example that can be added to this PR for the case in which we want to allow flash_read() to work but not flash_write()?

I feel like towards Andy's comment, that if we do allow for such an option, it would be DISABLE_WRITE_SUPPORT, and I'd think we'd make the API (flash_write) go away as part of it.

mbolivar commented:
Now that #654 has been approved, this should be closed.
